### DISCUSSION TITLE
Rasperry Pico + Waveshare Pico-LoRa-SX126X: Initializing ... failed, code -2

### DISCUSSION THREAD

**Discussion by jkarjanl** [jgromes/RadioLib#947] [Q&A] [answered]
Hello!

I'm willing to test LoRaWAN code on Rasperry Pico + Waveshare Pico-LoRa-SX126X HAT. However only output I can produce at the moment is 
`[SX1262] Initializing ... failed, code -2`

I tried LoRaWAN_End_Device.ino from examples. The Lora HAT is this exact model:
[https://www.waveshare.com/wiki/Pico-LoRa-SX1262](https://www.waveshare.com/wiki/Pico-LoRa-SX1262)

I'm not sure if I must adjust pin order from default? I tried different pins based Waveshare Wiki page, but it's always same.

Maybe I'm missing something trivial... or is this specific LoRaWAN Hat supported at all?
The board itself is working, I can produce some output to TTN cloud console using Waveshare's three year old (and then abandoned?) example code with official Raspberry Pico-SDK, so at least the hardware is functional.

**Comment by jgromes** [comment]
> I'm not sure if I must adjust pin order from default?

You have to provide the correct pinout. The library can run on any hardfware, how should it know which pins to use? Taht is up to you to configure. The pinout is on the page you linked. I don't know what is the default SPI the Pico uses, if the hat is using a different one you will have to configure it as well - see: https://github.com/jgromes/RadioLib/wiki/Basics#non-standard-spi-setup

>  is this specific LoRaWAN Hat supported at all?

The hat doesn't matter - it's SX1262 which is supported, and Raspberry Pico which is also supported.

> using Waveshare's three year old (and then abandoned?) example code with official Raspberry Pico-SDK

And which SDK are you using? The official Arduino one? The unofficial Arduino one? Or plain Pico SDK with some hardware abstraction layer? For the latter tehre is currently an open PR.

**Reply by jkarjanl** [reply]
Thank you for your clarification. I'm using official Arduino SDK, and because Pico and SX1262 both are supported I think the problem may be wrong pinout. I allready adjusted pinout based on Waweshare wiki pages and Raspberry pi pico GPIO layout.

// SX1262 has the following pin order:
// Module(NSS/CS, DIO1, RESET, BUSY)
SX1262 radio = new Module(5, 26, 20, 4);

However I'm not sure that I understand it correctly. I will investigate SPI configuration a little bit more.

**Reply by StevenCellist** [reply]
> https://www.waveshare.com/wiki/Pico-LoRa-SX1262

This link shows a nice image (https://waveshare.com/wiki/File:Pico-LoRa-SX1262-868M-details-inter.jpg). They clearly list CS = chip select, DIO1, RESET, BUSY...

How did you end up with these pin numbers??
> // Module(NSS/CS, DIO1, RESET, BUSY)
SX1262 radio = new Module(5, 26, 20, 4);

**Reply by jkarjanl** [reply]
CS, DIO1, RESET and BUSY looks like GP3, GP20, GP15 and GP2 using the picture, if I am correct?

In Raspberry Pi the usual way to use GPIO pins is using gpio.board style pin numbering, where GPIO numbers does not corresponde actual pin numbers, ie. it is using physical pin numbers. You can see the physical numbers of the pins in that picture as well. So GP3 is pin number 5, GP20 is pin number 26, GP15 is pin 20 and GP2 is pin number 4.

But... you made me thinking.. perhaps Arduino IDE is different world, and it uses gpio.bcm style numbering?
Unfortunately this still is not working:
`SX1262 radio = new Module(3, 20, 15, 2);`

I am not too familiar with SPI, and not familiar with Raspberry Pico at all, so maybe I didn't get all things right at all...

**Reply by StevenCellist** [reply]
> In Raspberry Pi the usual way to use GPIO pins is using gpio.board style pin numbering, where GPIO numbers does not corresponde actual pin numbers, ie. it is using physical pin numbers.

Ah, yeah, that makes sense, forgot that. Everything Arduino uses the GPIO pin numbering, which you got right this time around. However, the board you selected may not by default set the right SPI pins (sometimes it works, sometimes it doesn't). Personally, I use the radio initializer as follows:

```cpp
#include <SPI.h>

SPIClass spiSX(SPI);
SPISettings spiSettings(2000000, MSBFIRST, SPI_MODE0);
SX1262 radio = new Module(8, 14, 12, 13, spiSX, spiSettings);   // NSS/CS, DIO1, RESET, BUSY
```

And then in the `setup()`:
```cpp
spiSX.begin(9, 11, 10, 8);              // SCK/CLK, MISO, MOSI, NSS/CS
radio.begin();
```

Note that you will need to change your pin numbering for your board, this is an excerpt from my own code. Let us know if this helps or not!

**Reply by GUVWAF** [reply, marked as answer]
You have to use hardware SPI device 1 instead of 0 for this combination, see also: https://github.com/jgromes/RadioLib/issues/729#issuecomment-1582304326

`SX1262 radio = new Module(3, 20, 15, 2, SPI1, RADIOLIB_DEFAULT_SPI_SETTINGS);` are the correct pins.

**Reply by jkarjanl** [reply]
Thank you all, only high quality answers here :-)

I can confirm that @GUVWAF was right on the spot here.

`SX1262 radio = new Module(3, 20, 15, 2, SPI1, RADIOLIB_DEFAULT_SPI_SETTINGS);`

However compiler throw errors with SPI1 when using official "Arduino Mbed OS RP2040 Boards" -core. I changed to earlephilhowers "Arduino-Pico" -core and after that it compiles ok. The core is this one:
[https://github.com/earlephilhower/arduino-pico](https://github.com/earlephilhower/arduino-pico)

Unfortunately result was still same:
[SX1262] Initializing ... failed, code -2

So I added GUVWAF's suggestions from another thread to the beginning of setup() -function, and now it's working!
```
#include <RadioLib.h>
// SX1262 has the following pin order:
// Module(NSS/CS, DIO1, RESET, BUSY)
//SX1262 radio = new Module(3, 20, 15, 2);
SX1262 radio = new Module(3, 20, 15, 2, SPI1, RADIOLIB_DEFAULT_SPI_SETTINGS);

// create the node instance on the EU-868 band
// using the radio module and the encryption key
// make sure you are using the correct band
// based on your geographical location!
LoRaWANNode node(&radio, &EU868);

void setup() {
  Serial.begin(9600);
  delay(2000);

  // from GUVWAF's post:
  SPI1.setSCK(10);
  SPI1.setTX(11);
  SPI1.setRX(12);
  pinMode(3, OUTPUT);
  digitalWrite(3, HIGH);
  SPI1.begin(false);

  // initialize SX1262 with default settings
  Serial.print(F("[SX1262] Initializing ... "));
  int state = radio.begin();
---snip---
```
So the ouput looks like this:
[SX1262] Initializing ... success!
[LoRaWAN] Attempting over-the-air activation ... failed, code -6

I can see join request in TTN Console, so the data is definitely flowing from my livingroom in Finland to TTN cloud located in Ireland. I am tempted to call it success, altought the join request itself fails with "MIC mismatch" -error. I think it's related DevEUI and JoinEUI and all that boring stuff, hopefully I can manage that myself.

Thank you.

**Reply by LucasWolf-Pudney** [reply]
Hi, what example did you base your code off?  I am using the LoRaWAN_starter with the same module you are using, and changing my pins in the config and I get this error. 
Setup ... 
Initialise the radio
Join ('login') the LoRaWAN Network
Join failed - RADIOLIB_ERR_NO_JOIN_ACCEPT (-1116)
I think I am getting the sames kind of issues you had where there is incroecct join EUI and other paramteres like that.
